### ISSUE TITLE
Field validator as method is buggy/inconsistent

### ISSUE THREAD

**Issue by martinstein** [marshmallow-code/marshmallow#391]
Let's take the example from the docs at http://marshmallow.readthedocs.org/en/latest/quickstart.html#field-validators-as-methods

If you run this:

``` python
from marshmallow import Schema, fields, validates, ValidationError

class ItemSchema(Schema):
    quantity = fields.Integer()

    @validates('quantity')
    def validate_quantity(self, value):
        if value < 0:
            raise ValidationError('Quantity must be greater than 0.')
        if value > 30:
            raise ValidationError('Quantity must not be greater than 30.')

schema = ItemSchema()
result, errors = schema.load({'quantity': 40})
print("result:", result) 
print("errors:", errors)
```

you get the result:

```
result: {'quantity': 40}
errors: {'quantity': ['Quantity must not be greater than 30.']}
```

instead of the empty result `{}` (because validation failed).

Just for comparison, the behavior is correct for validation via lambda-function:

``` python
class ItemSchema2(Schema):
    quantity = fields.Integer(validate=lambda x: 0 <= x <= 30)

schema = ItemSchema2()
result, errors = schema.load({'quantity': 40})
print("result:", result)
print("errors:", errors)
```

This time you get the expected result:

```
result: {}
errors: {'quantity': ['Invalid value.']}
```

**Comment by sloria** [comment]
Thanks for reporting @martinstein . I would certainly review and merge a PR addressing this inconsistency.

**Comment by jmcarp** [comment]
It looks like this inconsistency might get fixed for free once #368 is merged and `validates` is deprecated.

**Comment by vuonghv** [comment]
@sloria @jmcarp 
Do you think we really need a PR to fix this issue because `validates` is deprecated in the future (#368)?

**Comment by sloria** [comment]
Even though `validates` will be deprecated after #368, I would gladly review and merge a PR fixing the inconsistency.

**Comment by maximkulkin** [comment]
Isn't that if validation fails, the contents of result does not make sense and could just be garbage/inconsistent? Where can I find documentation on guarantees of what result will be in case of errors?

**Comment by vuonghv** [comment]
I think that removing the invalid values from `result` will make getting all valid values easier by iterating `result`'s keys. We do not have to check if the key is in `errors` or not.

**Comment by maximkulkin** [comment]
Valid values should be determined by absence of errors for that path. How will removing fields help detecting good values? Do you need ANOTHER validator to distinguish values that have all fields from values that do not have all fields?

Also, what do you do with those valid values in the event of invalid values presence? Are you making this up or there is a good use case for that?

**Comment by vuonghv** [comment]
@maximkulkin Lets come back to the main purpose of this issue. We want to be consistent in returning `result` and `errors`, regardless of the ways we use validators (`validates` decorator or pass function to the `validate` param in Field's `__init__`).
Do you think it is a good idea?

**Comment by maximkulkin** [comment]
I just think there is no value in being consistent in uselessness and doing anything for that matter.

**Comment by martinstein** [comment]
@vuonghv Thanks for providing this pull request. Fixing this inconsistency would be great.

@maximkulkin If you disagree with the original behavior of removing invalid values from results, I think that should be part of a separate issue/discussion. vuonghv's pull request is a clear improvement since it makes marhsmallow's behavior more consistent. Let's not fall into the trap where we ignore improvements in one area just because there might be disagreement in another area.

So I think this is definitely good to be merged.

**Comment by maximkulkin** [comment]
Please then mark this as backward incompatible as it changes current behavior.

**Comment by sloria** [comment]
This is a bug, since the expected behavior is that `@validates` is consistent with field validators. Therefore, the PR is considered a fix rather than a breaking change.
